### PR TITLE
Adds separable convolution block

### PR DESCRIPTION
**Pull request by bedapisl**

### CODE REVIEW

File: emloop_tensorflow/models/conv_blocks.py [hunk 1]
```diff
@@ -164,6 +166,52 @@ def inverse_code(self) -> str:
...
+        :param stride: stride (default 1)
+        """
+        self._channels, self._kernel, self._stride = int(channels), int(kernel), int(stride)
+
+    def apply(self, x: tf.Tensor) -> tf.Tensor:
+        if len(self._extra_dim) >= 1:
+            raise ValueError('SeparableConvBlock only supports inputs with rank 4 \
+                              (i.e. batch_size, height, width, channels)')
+
+        x = slim.separable_conv2d(x, num_outputs=self._channels, kernel_size=(self._kernel, self._kernel),
```
Comment (FloopCZ) [inline emloop_tensorflow/models/conv_blocks.py]: (I have to admit that I would prefer `tf.layers` in new code, but the variable scopes would not work. :-/)
https://www.tensorflow.org/api_docs/python/tf/layers/separable_conv2d

File: emloop_tensorflow/models/conv_blocks.py [hunk 2]
```diff
@@ -164,6 +166,52 @@ def inverse_code(self) -> str:
...
+    """
+
+    def __init__(self, **kwargs):
+        """Try to parse and create new :py:class:`SeparableConvBlock`."""
+        super().__init__(regexp='([1-9][0-9]*)sep([1-9][0-9]*)(s([1-9][0-9]*))?',
+                         defaults=(None, None, None, 1),
+                         **kwargs)
+
+    def _handle_parsed_args(self, channels: str, kernel: str,
+                            __, stride: Union[str, int]) -> None:
```
Comment (FloopCZ) [inline emloop_tensorflow/models/conv_blocks.py]: Double underscore.

File: emloop_tensorflow/models/conv_blocks.py [hunk 3]
```diff
@@ -164,6 +166,52 @@ def inverse_code(self) -> str:
...
 
+class SeparableConvBlock(ConvBaseBlock):
+    """
+    2D depthwise separable convolutional block, as described in
+    'MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications'
+    (https://arxiv.org/pdf/1704.04861.pdf)
+
+    **code**: ``(num_filters)sep(kernel_size)[s(stride)]``
+
+    **examples**: ``64sep3``, ``64sep3s2``, ``64sep3-5s2``
```
Comment (FloopCZ) [inline emloop_tensorflow/models/conv_blocks.py]: Does `64sep3-5s2` actually work? :slightly_smiling_face:

File: emloop_tensorflow/models/conv_blocks.py [hunk 4]
```diff
@@ -164,6 +166,52 @@ def inverse_code(self) -> str:
...
+
+        :param channels: number of output channels
+        :param kernel: kernel size
+        :param _: void parameter, needed because function signature must match the groups in regexp from constructor
+        :param stride: stride (default 1)
+        """
+        self._channels, self._kernel, self._stride = int(channels), int(kernel), int(stride)
+
+    def apply(self, x: tf.Tensor) -> tf.Tensor:
+        if len(self._extra_dim) >= 1:
```
Comment (FloopCZ) [inline emloop_tensorflow/models/conv_blocks.py]: Why not check the length of the shape of `x` directly?

File: emloop_tensorflow/models/conv_blocks.py [hunk 5]
```diff
@@ -164,6 +166,52 @@ def inverse_code(self) -> str:
...
+                              (i.e. batch_size, height, width, channels)')
+
+        x = slim.separable_conv2d(x, num_outputs=self._channels, kernel_size=(self._kernel, self._kernel),
+                                  stride=(self._stride, self._stride), scope='inner')
+
+        return self._ln_fn(self._bn_fn(x))
+
+    def inverse_code(self) -> str:
+        if self._stride > 1:
+            raise ValueError('Inverse code for conv block is not defined for stride `{}`'.format(self._stride))
```
Comment (FloopCZ) [inline emloop_tensorflow/models/conv_blocks.py]: separable conv block

File: emloop_tensorflow/models/conv_blocks.py (hunk 3 above)
Comment (bedapisl) [reply emloop_tensorflow/models/conv_blocks.py]: No, my fault.

**Pull request merged**